provider/kubeletconfig/kubeletconfig_resource.go

```diff
@@ -0,0 +1,257 @@
+}
+
+func (k *KubeletConfigResource) getKubeletConfigStateFromState(
+	ctx context.Context, state tfsdk.State, diags *diag.Diagnostics) *KubeletConfigState {
+	ks := &KubeletConfigState{}
+	ds := state.Get(ctx, ks)
+	diags.Append(ds...)
+	return ks
+}
+
+func (k *KubeletConfigResource) getKubeletConfigStateFromPlan(
+	ctx context.Context, plan tfsdk.Plan, diags *diag.Diagnostics) *KubeletConfigState {
+	ks := &KubeletConfigState{}
+	ds := plan.Get(ctx, ks)
+	diags.Append(ds...)
+	return ks
+}
+
+func (k *KubeletConfigResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	plan := k.getKubeletConfigStateFromState(ctx, req.State, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	clusterId := plan.Cluster.ValueString()
+	exists, kubeletConfig, err := k.configClient.Exists(ctx, clusterId)
+	if err != nil {
+		resp.Diagnostics.AddError("Cannot read KubeletConfig",
+			fmt.Sprintf("Cannot read KubeletConfig for cluster '%s': %v", clusterId, err))
+		return
+	}
+
+	if !exists {
+		tflog.Warn(ctx, fmt.Sprintf("kubeletconfig for cluster (%s) not found, removing from plan",
+			plan.Cluster.ValueString(),
+		))
+		resp.State.RemoveResource(ctx)
+		return
+	}
+
+	k.convertApiResourceToState(kubeletConfig, plan)
+	k.writeStateToResponse(ctx, plan, &resp.State, &resp.Diagnostics)
+}
+
+func (k *KubeletConfigResource) writeStateToResponse(
+	ctx context.Context, ks *KubeletConfigState, tfState *tfsdk.State, diagnostics *diag.Diagnostics) {
+
+	diags := tfState.Set(ctx, ks)
+	diagnostics.Append(diags...)
+}
+
+func (k *KubeletConfigResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+
+	plan := k.getKubeletConfigStateFromPlan(ctx, req.Plan, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	clusterId := plan.Cluster.ValueString()
+
+	if _, err := k.configClient.Get(ctx, clusterId); err != nil {
+		resp.Diagnostics.AddError(failedToUpdateSummary,
+			fmt.Sprintf("Failed to read details of existing KubeletConfig for cluster '%s': %v", clusterId, err))
+		return
+	}
+
+	kubeletConfig, err := k.convertStateToApiResource(plan)
+	if err != nil {
+		resp.Diagnostics.AddError(failedToUpdateSummary,
+			fmt.Sprintf("Failed to build request to update existing KubeletConfig for cluster '%s': %v", clusterId, err))
+		return
+	}
+
+	updateResponse, err := k.configClient.Update(ctx, clusterId, kubeletConfig)
+	if err != nil {
+		resp.Diagnostics.AddError(failedToUpdateSummary,
+			fmt.Sprintf("Failed to update existing KubeletConfig for cluster '%s': %v", clusterId, err))
+		return
+	}
+
+	k.convertApiResourceToState(updateResponse, plan)
+	k.writeStateToResponse(ctx, plan, &resp.State, &resp.Diagnostics)
+}
+
+func (k *KubeletConfigResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	plan := k.getKubeletConfigStateFromState(ctx, req.State, &resp.Diagnostics)
```

Comment:
Why did you change `state` to `plan` here
The comment wasn't relevant to here
Here you are using the `state` (and not the `plan`)

Reply:
I fixed this and going to push additional commit to this PR, to unblock the merge